desktop-widgets/templatelayout.cpp

```diff
@@ -286,35 +284,47 @@ static QRegularExpression forloop(R"(\s*(\w+)\s+in\s+(\w+))");	// Look for "VAR
 static QRegularExpression ifstatement(R"(forloop\.counter\|\s*divisibleby\:\s*(\d+))");	// Look for forloop.counter|divisibleby: NUMBER
 
 template<typename V, typename T>
-void TemplateLayout::parser_for(QList<token> tokenList, int &pos, QTextStream &out, State &state,
+void TemplateLayout::parser_for(QList<token> tokenList, int from, int to, QTextStream &out, State &state,
 				const V &data, const T *&act)
 {
 	const T *old = act;
 	int i = 1; // Loop iterators start at one
 	int olditerator = state.forloopiterator;
-	int savepos = pos;
 	for (const T &item: data) {
 		act = &item;
-		state.forloopiterator = i;
-		pos = savepos;
-		++i;
-		parser(tokenList, pos, out, state);
+		state.forloopiterator = i++;
+		parser(tokenList, from, to, out, state);
 	}
 	act = old;
 	state.forloopiterator = olditerator;
 }
 
-void TemplateLayout::parser(QList<token> tokenList, int &pos, QTextStream &out, State &state)
```

Comment:
Implementing this in this  more procedural look ahead way rather than the functional one I had implemented is a matter of personal preference I would say. Also the functional approach could produce better error messages by testing for this in the handling of endfor/endif etc. But this definitely works, so fine with me.

Reply:
Note that it wasn't only bad diagnostics, but - if I read it correctly - the old code was buggy for empty lists - which can happen for cylinders. If the loop body was not executed once, `pos` was not increased accordingly, isn't it?

Not trying to get into pointless semantic discussions, but the old code was *not* functional, and that was why it was hard to reason about. The whole advantage (some may argue disadvantage) of functional programming is that functions have no side effects. For the same input a functions *must* give the same output and it *must never* change the input parameters. Therefore, the line that you quote is precisely an example of what functional programming is not. For another example, `i = i + 1` makes no sense in functional programming.

That said, the correct way would of course be constructing a parse-tree, not a list of tokens. However, for now that's probably not worth it.